lib/state-store.js

```diff
@@ -0,0 +1,311 @@
+      awsOptions,
+      consumerGroup,
+      consumerId,
+      logger,
+      streamName,
+      tableName: tableName || `${name}-state`,
+      tags
+    });
+  }
+
+  /**
+   * Starts the state store by initializing a DynamoDB client and a document client. Then,
+   * it will ensure the table exists, that is tagged as required, and there's an entry for
+   * the stream state.
+   */
+  async start() {
+    const privateProps = internal(this);
+    const { awsOptions, consumerGroup, logger, streamName, tableName, tags } = privateProps;
+
+    const client = new DynamoDBProxy({ awsOptions, tableName });
+    privateProps.client = client;
+
+    let tableArn = await checkIfTableExists(client, logger, tableName);
+    if (!tableArn) tableArn = await createTable(client, logger, tableName);
+    await confirmTableTags(client, logger, tableArn, tags);
+
+    try {
+      const params = {
+        ConditionExpression: 'attribute_not_exists(streamName)',
+        Item: {
+          consumerGroup,
+          consumers: {},
+          shards: {},
+          streamName,
+          version: generate()
+        }
+      };
+      await client.put(params);
+      logger.debug('Initial state has been recorded for the stream.');
+    } catch (err) {
+      if (err.code !== 'ConditionalCheckFailedException') {
+        logger.error(err);
+        throw err;
+      }
+    }
+  }
+
+  async clearOldConsumers(heartbeatFailureTimeout) {
+    const privateProps = internal(this);
+    const { consumerGroup, client, logger, streamName } = privateProps;
+
+    const { consumers, version } = await getStreamState(this);
+    const consumerIds = Object.keys(consumers);
+
+    const oldConsumers = consumerIds.filter(id => {
+      const { heartbeat } = consumers[id];
+      return Date.now() - new Date(heartbeat).getTime() > heartbeatFailureTimeout;
+    });
+
+    if (oldConsumers.length === 0) return;
+
+    const params = {
+      Key: { consumerGroup, streamName },
+      UpdateExpression: `REMOVE ${oldConsumers
+        .map((id, index) => `#a.#${index}`)
+        .join(', ')} SET #b = :x`,
+      ConditionExpression: `#b = :y`,
+      ExpressionAttributeNames: {
+        '#a': 'consumers',
+        '#b': 'version',
+        ...oldConsumers.reduce((obj, id, index) => ({ ...obj, [`#${index}`]: id }), {})
+      },
+      ExpressionAttributeValues: {
+        ':x': generate(),
+        ':y': version
+      }
+    };
+
+    try {
+      await client.update(params);
+      logger.debug(`Cleared old consumers: ${oldConsumers.join(', ')}`);
+    } catch (err) {
+      if (err.code !== 'ConditionalCheckFailedException') {
+        logger.error(err);
+        throw err;
+      }
+      logger.debug('Old consumers were cleared somewhere else.');
+    }
+  }
+
+  async registerConsumer() {
+    const privateProps = internal(this);
+    const { consumerId, consumerGroup, client, logger, streamName } = privateProps;
+
+    const heartbeat = new Date().toISOString();
```

Comment:
What does this "heartbeat" variable represent?

Reply:
The current timestamp in ISO format.